cpp/src/text/subword/data_normalizer.cu

```diff
@@ -0,0 +1,459 @@
+ */
+__device__ __forceinline__ uint32_t get_first_cp(uint32_t metadata)
+{
+  return metadata & NEW_CP_MASK;
+}
+
+/**
+ * @brief Retrieve token category from the metadata value.
+ *
+ * Category values are 0-5:
+ * 0 - character should be padded
+ * 1 - pad character if lower-case
+ * 2 - character should be removed
+ * 3 - remove character if lower-case
+ * 4 - whitespace character -- always replace
+ * 5 - uncategorized
+ *
+ * @param metadata Value from the codepoint_metadata table.
+ * @return Category value.
+ */
+__device__ __forceinline__ uint32_t extract_token_cat(uint32_t metadata)
+{
+  return (metadata >> TOKEN_CAT_SHIFT) & TOKEN_CAT_MASK;
+}
+
+/**
+ * @brief Return true if category of metadata value specifies the character should be replaced.
+ */
+__device__ __forceinline__ bool should_remove_cp(uint32_t metadata, bool lower_case)
+{
+  auto const cat = extract_token_cat(metadata);
+  return (cat == TOKEN_CAT_REMOVE_CHAR) || (lower_case && (cat == TOKEN_CAT_REMOVE_CHAR_IF_LOWER));
+}
+
+/**
+ * @brief Return true if category of metadata value specifies the character should be padded.
+ */
+__device__ __forceinline__ bool should_add_spaces(uint32_t metadata, bool lower_case)
+{
+  auto const cat = extract_token_cat(metadata);
+  return (cat == TOKEN_CAT_ADD_SPACE) || (lower_case && (cat == TOKEN_CAT_ADD_SPACE_IF_LOWER));
+}
+
+/**
+ * @brief Return true if category of metadata value specifies the character should be replaced.
+ */
+__device__ __forceinline__ bool always_replace(uint32_t metadata)
+{
+  return extract_token_cat(metadata) == TOKEN_CAT_ALWAYS_REPLACE;
+}
+
+/**
+ * @brief Returns true if metadata value includes a multi-character transform bit equal to 1.
+ */
+__device__ __forceinline__ bool is_multi_char_transform(uint32_t metadata)
+{
+  return (metadata >> MULTICHAR_SHIFT) & MULTICHAR_MASK;
+}
+
+/**
+ * @brief Returns true if the byte passed in could be a valid head byte for
+ * a utf8 character. That is, not binary `10xxxxxx`
+ */
+__device__ __forceinline__ bool is_head_byte(unsigned char utf8_byte)
+{
+  return (utf8_byte >> 6) != 2;
+}
+
+/**
+ * @brief Converts a UTF-8 character into a unicode code point value.
+ *
+ * If the byte at start_byte_for_thread is the first byte of a UTF-8 character (head byte),
+ * the UTF-8 character is converted to a unicode code point and returned.
+ *
+ * If the byte at start_byte_for_thread is not a head byte, 0 is returned.
+ *
+ * All threads start reading bytes from the pointer denoted by strings.
+ *
+ * @param strings A pointer to the start of the sequence of characters to be analyzed.
+ * @param start_byte_for_thread Which byte to start analyzing
+ * @return New code point value for this byte.
+ */
+__device__ __forceinline__ uint32_t
+extract_code_points_from_utf8(const unsigned char* strings, const uint32_t start_byte_for_thread)
+{
+  constexpr uint8_t max_utf8_blocks_for_char = 4;
+  uint8_t utf8_blocks[max_utf8_blocks_for_char];
+
+#pragma unroll
```

Comment:
BTW, do you ensure that the size of strings is a multiple of 4 bytes? Otherwise you will be accessing out of bounds.

Reply:
> You could replace this whole loop with a single 32-bit load with appropriate casts. That would be the most efficient result (and would be nice if the compiler does it automatically)...

These values are not aligned on 4-byte values when the kernel is executed. There is a thread per byte so this would require some extra logic. Regardless, I believe the individual bytes are retrieved from L1 cache when not aligned.
I think this function could be reworked in a follow on PR.